File: 2015/05/main.js
const fs = require("node:fs")
const path = require("node:path")

const filename = "input"
const file =
    fs.readFileSync(
        path.resolve(__dirname, filename),
        {
            encoding: "utf-8"
        }
    )
    .trim()
    .split("\n")

// const vowels = "aeiou".split("")
//
// function duplicatedVowel(line, vowels) {
//     const countVowelsObject = {}
//     for (let letter of line) {
//         if(vowels.includes(letter)) {
//             if (countVowelsObject[letter]) {
//                 countVowelsObject[letter]++
//             } else {
//                 countVowelsObject[letter] = 1
//             }
//         }
//     }
//     return Object.values(countVowelsObject)
//         .reduce((prev, curr) => prev + curr, 0) >= 3
// }
//
// function notAllowedLetters(line) {
//     return !["ab", "cd", "pq", "xy"].some(pair => line.includes(pair))
// }
//
// function createDoubleAlphabet() {
//     let alphabet = ""
//     for (let i = 65; i <= 90; i++) {
//         alphabet = alphabet + String.fromCharCode(i).toLowerCase()
//     }
//     const double_alpha = alphabet
//         .split("")
//         .map(char => char + char)
//     return double_alpha
// }
//
// function hasDoubleAlphaLetter(line, double_alpha) {
//     for (let double of double_alpha) {
//         if (line.includes(double)) {
//             return true
//         }
//     }
//     return false
// }
//     
// const double_alpha = createDoubleAlphabet()
// const file_with_allowed = file
//     .filter(notAllowedLetters)
//     .filter(line => hasDoubleAlphaLetter(line, double_alpha))
//     .filter(line => duplicatedVowel(line, vowels))

// Code Reviewed part_one
const vowels = new Set("aeiou") // <-- O(1)

function hasThreeVowels(line) {
    let vowelCount = 0
    for (const char of line) {
        if (vowels.has(char)) {
            vowelCount++
        }
        if (vowelCount >= 3) return true
    }
    return false
}

function noForbiddenSubstrings(line) {
    const forbidden = ["ab", "cd", "pq", "xy"]
    return forbidden.every(pair => !line.includes(pair))
}

function hasDoubleLetter(line) {
    for (let i = 0; i < line.length - 1; i++) {
        if (line[i] === line[i + 1]) {
            return true
        }
    }
    return false
}

const file_with_allowed = file
    .filter(noForbiddenSubstrings)
    .filter(hasDoubleLetter)
    .filter(hasThreeVowels)

console.log({ 
    len: file.length,
    new_len: file_with_allowed.length,
})

File: 2015/05/main.exs
defmodule Utils do 
  @spec read_file(String.t()) :: String.t()
  defp read_file(filename) do
    filename
    |> File.read()
    |> case do
      {:ok, file} -> 
        file
        |> String.trim()
      _ -> IO.puts("error reading file {filename}")
    end
  end

  @spec split_file_line(String.t()) :: list(String.t())
  defp split_file_line(file) do
    file
    |> String.split("\n")
  end
end

defmodule PartOne do
  @spec create_vowels() :: list(String.t())
  defp create_vowels() do
    "aeiou"
      |> String.graphemes()
  end

  @spec no_forbiden_substrings?(String.t()) :: Atom.t()
  defp no_forbiden_substrings?(line) do
    not (
      line
      |> String.contains?(["ab", "cd", "pq", "xy"])
    )
  end

  @spec has_three_vowels?(String.t(), list(String.t())) :: Atom.t()
  defp has_three_vowels?(line, vowels) do
    line
    |> String.graphemes()
    |> Enum.reduce(0, fn letter, acc -> 
      case Enum.member?(vowels, letter) do
        true -> acc + 1
        _ -> acc
      end
    end)
    |> case do
        x when x >= 3 -> true
        _ -> false
      end
  end

  @spec has_double_alpha_letter?(String.t()) :: Atom.t()
  defp has_double_alpha_letter?(line) do
    line
    |> String.graphemes()
    |> Enum.chunk_by(fn arg -> arg end)
    |> Enum.map(fn arg -> to_string(arg) end)
    |> Enum.filter(fn x -> String.length(x) > 1 end)
    |> length()
    |> case do
      x when x > 0 -> true
      _ -> false
    end
  end

  @spec resolve(String.t()) :: Integer.t()
  def resolve(filename) do
    vowels = create_vowels()
    filename
      |> Utils.read_file()
      |> Utils.split_file_line()
      |> Enum.filter(fn line ->
          no_forbiden_substrings?(line)
          and has_double_alpha_letter?(line)
          and has_three_vowels?(line, vowels)
      end)
      |> length()
  end
end

filename = "input"
Resolver.part_one(filename)
  |> IO.inspect()

